File: runtime/Swift/Sources/Antlr4/atn/AtomTransition.swift
/// Copyright (c) 2012-2016 The ANTLR Project. All rights reserved.
/// Use of this file is governed by the BSD 3-clause license that
/// can be found in the LICENSE.txt file in the project root.


/// TODO: make all transitions sets? no, should remove set edges

public final class AtomTransition: Transition, CustomStringConvertible {
    /// The token type or character value; or, signifies special label.
    public let label: Int

    public init(_ target: ATNState, _ label: Int) {

        self.label = label
        super.init(target)
    }

    override
    public func getSerializationType() -> Int {
        return Transition.ATOM
    }

    override
    public func labelIntervalSet() throws -> IntervalSet? {
        return try IntervalSet.of(label)
    }

    override
    public func matches(_ symbol: Int, _ minVocabSymbol: Int, _ maxVocabSymbol: Int) -> Bool {
        return label == symbol
    }


    public var description: String {
        return String(label)
    }
}

File: runtime/JavaScript/src/antlr4/transition/AtomTransition.js
/* Copyright (c) 2012-2022 The ANTLR Project Contributors. All rights reserved.
 * Use is of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */
import IntervalSet from "../misc/IntervalSet.js";
import Transition from "./Transition.js";

export default class AtomTransition extends Transition {
    constructor(target, label) {
        super(target);
        // The token type or character value; or, signifies special label.
        this.label_ = label;
        this.label = this.makeLabel();
        this.serializationType = Transition.ATOM;
    }

    makeLabel() {
        const s = new IntervalSet();
        s.addOne(this.label_);
        return s;
    }

    matches(symbol, minVocabSymbol, maxVocabSymbol) {
        return this.label_ === symbol;
    }

    toString() {
        return this.label_;
    }
}
